File: src/main/java/io/github/manhdua1/lotusoj/service/testCase/TestCaseService.java
package io.github.manhdua1.lotusoj.service.testCase;

import io.github.manhdua1.lotusoj.dto.request.testCase.CreateTestCaseRequest;
import io.github.manhdua1.lotusoj.dto.request.testCase.UpdateTestCaseRequest;
import io.github.manhdua1.lotusoj.dto.response.testCase.TestCaseResponse;

import java.util.List;
import java.util.UUID;

public interface TestCaseService {

    /**
     * Lấy tất cả test case của một bài (sắp xếp theo orderIndex tăng dần)
     * để phục vụ cho việc chấm bài.
     */
    List<TestCaseResponse> getTestCasesForJudging(UUID problemId);

    /**
     * Lấy chỉ các test case mẫu (isSample = true) để hiển thị cho người dùng.
     */
    List<TestCaseResponse> getSampleTestCases(UUID problemId);

    /**
     * Lấy tất cả test cases của một problem theo thứ tự orderIndex (dành cho Admin / Problem Setter).
     */
    List<TestCaseResponse> getTestCasesByProblemId(UUID problemId);

    /**
     * Tạo một test case mới gán cho Problem.
     */
    TestCaseResponse createTestCase(UUID problemId, CreateTestCaseRequest request);

    /**
     * Cập nhật thông tin/nội dung của một test case.
     */
    TestCaseResponse updateTestCase(UUID testCaseId, UpdateTestCaseRequest request);

    /**
     * Xóa một test case khỏi hệ thống.
     */
    void deleteTestCase(UUID testCaseId);

    /**
     * Đếm tổng số test case của một bài.
     */
    long countByProblemId(UUID problemId);
}

File: frontend/src/services/testCaseService.ts
import type { ApiResponse } from '../types/problem'
import type { CreateTestCaseRequest, TestCaseResponse, UpdateTestCaseRequest } from '../types/testCase'
import { authService } from './authService'

export const testCaseService = {
  /**
   * Lấy danh sách testcase mẫu (isSample = true) công khai
   */
  async getSampleTestCases(problemId: string): Promise<TestCaseResponse[]> {
    const response = await authService.fetchWithAuth(`/api/problems/${problemId}/testcases/samples`, {
      method: 'GET',
    })

    if (!response.ok) {
      let errorMsg = `Lỗi tải danh sách test case mẫu (HTTP ${response.status})`
      try {
        const errJson: ApiResponse<unknown> = await response.json()
        if (errJson.message) errorMsg = errJson.message
      } catch {
        // fallback
      }
      throw new Error(errorMsg)
    }

    const data: ApiResponse<TestCaseResponse[]> = await response.json()
    if ((data.code !== 1000 && data.code !== 200) || !data.result) {
      throw new Error(data.message || 'Lỗi xử lý yêu cầu test case')
    }

    return data.result
  },

  /**
   * Lấy toàn bộ danh sách testcase của bài tập (chỉ dành cho Admin / Problem Setter)
   */
  async getProblemTestCases(problemId: string): Promise<TestCaseResponse[]> {
    const response = await authService.fetchWithAuth(`/api/problems/${problemId}/testcases`, {
      method: 'GET',
    })

    if (!response.ok) {
      let errorMsg = `Lỗi tải bộ test của bài tập (HTTP ${response.status})`
      try {
        const errJson: ApiResponse<unknown> = await response.json()
        if (errJson.message) errorMsg = errJson.message
      } catch {
        // fallback
      }
      throw new Error(errorMsg)
    }

    const data: ApiResponse<TestCaseResponse[]> = await response.json()
    if ((data.code !== 1000 && data.code !== 200) || !data.result) {
      throw new Error(data.message || 'Không thể lấy danh sách test case')
    }

    return data.result
  },

  /**
   * Tạo 1 testcase mới cho bài tập
   */
  async createTestCase(problemId: string, request: CreateTestCaseRequest): Promise<TestCaseResponse> {
    const response = await authService.fetchWithAuth(`/api/problems/${problemId}/testcases`, {
      method: 'POST',
      body: JSON.stringify(request),
    })

    if (!response.ok) {
      let errorMsg = `Lỗi tạo test case (HTTP ${response.status})`
      try {
        const errJson: ApiResponse<unknown> = await response.json()
        if (errJson.message) errorMsg = errJson.message
      } catch {
        // fallback
      }
      throw new Error(errorMsg)
    }

    const data: ApiResponse<TestCaseResponse> = await response.json()
    if ((data.code !== 1000 && data.code !== 200 && data.code !== 201) || !data.result) {
      throw new Error(data.message || 'Không thể tạo test case mới')
    }

    return data.result
  },

  /**
   * Cập nhật 1 testcase
   */
  async updateTestCase(testCaseId: string, request: UpdateTestCaseRequest): Promise<TestCaseResponse> {
    const response = await authService.fetchWithAuth(`/api/testcases/${testCaseId}`, {
      method: 'PUT',
      body: JSON.stringify(request),
    })

    if (!response.ok) {
      let errorMsg = `Lỗi cập nhật test case (HTTP ${response.status})`
      try {
        const errJson: ApiResponse<unknown> = await response.json()
        if (errJson.message) errorMsg = errJson.message
      } catch {
        // fallback
      }
      throw new Error(errorMsg)
    }

    const data: ApiResponse<TestCaseResponse> = await response.json()
    if ((data.code !== 1000 && data.code !== 200) || !data.result) {
      throw new Error(data.message || 'Không thể cập nhật test case')
    }

    return data.result
  },

  /**
   * Xóa 1 testcase
   */
  async deleteTestCase(testCaseId: string): Promise<void> {
    const response = await authService.fetchWithAuth(`/api/testcases/${testCaseId}`, {
      method: 'DELETE',
    })

    if (!response.ok) {
      let errorMsg = `Lỗi xóa test case (HTTP ${response.status})`
      try {
        const errJson: ApiResponse<unknown> = await response.json()
        if (errJson.message) errorMsg = errJson.message
      } catch {
        // fallback
      }
      throw new Error(errorMsg)
    }
  },
}
